_overviews/scala3-book/domain-modeling-oop.md

```diff
@@ -0,0 +1,241 @@
+p.removeTopping(Pepperoni)
+
+// print the pizza, which uses its `toString` method
+println(p)
+```
+
+That `println` statement prints this output:
+
+````
+Pizza:
+  Crust Size: Large
+  Crust Type: Thick
+  Toppings:   ArrayBuffer(Cheese, BlackOlives)
+````
+
+Hopefully if you’re familiar with OOP languages like Java and Python, this code looks similar to what you’re used to.
+
+
+
+
+### Using traits as interfaces
+
+In object-oriented programming it’s also common to use interfaces, and then implement those interfaces with classes. When writing OOP code with Scala you do the same thing: you create interfaces as traits, and then create classes that extend those traits.
+
+For example, this is a trait that defines an attribute named `numLegs`, and two behaviors. The `walk()` behavior is abstract, because it doesn’t have a method body, and `stop()` is a concrete method, because it has a method body. Even though `stop()` is defined, it can also be overridden in your classes:
+
+```scala
+trait HasLegs:
+  def numLegs: Int
+  def walk(): Unit
+  def stop() = println("Stopped walking")
+```
+
+Similarly, this is a trait that defines two abstract behaviors for a “tail,” such as a tail on a dog, cat, or other animals:
+
+```scala
+trait HasTail:
+  def wagTail(): Unit
+  def stopTail(): Unit
+```
+
+Now you can define a class named `Dog` that extends both of those traits, and implements their behaviors:
+
+```scala
+class Dog(var name: String) extends HasLegs, HasTail:
+  val numLegs = 4
+  def walk() = println("I’m walking")
+  def wagTail() = println("⎞⎜⎛  ⎞⎜⎛")
+  def stopTail() = println("Tail is stopped")
+  override def toString = s"$name is a Dog"
+
+// create a Dog instance
+val d = Dog("Rover")
+
+// use the attributes and behaviors
+println(d.numLegs)   // 4
+d.wagTail()          // "⎞⎜⎛  ⎞⎜⎛"
+d.walk()             // "I’m walking"
+```
+
+Note that the methods `walk()`, `wagTail()`, and `stopTail()` are defined with empty parentheses. This isn’t necessary, you could define them as `walk`, `wagTail`, and `stopTail`, but a Scala recommendation is to name methods like these that have *side effects* with open parentheses. When you do this, it requires callers of these methods to use open parentheses when they call the methods, and this is a reminder that these methods have side effects.
+<!-- TODO: link to a "side effects" discussion -->
+
+Scala classes have other features to support OOP-style programming:
+
+- You can define methods and fields to be `private` or `protected`, to control how they are accessed and overridden
+- Classes can have multiple constructors
+- As shown in the [Modeling Tools section](domain-modeling-tools.md), class constructor parameters can have default values
+- A class can be marked as `open`, signaling that it’s planned for extensions:
+
+```scala
+open class MyOpenClass:
+  // ...
+```
+
+For more details on working with classes, see the Reference documentation.
+<!-- TODO: link to the Reference docs -->
+
+
```

Comment:
When talking about OOP in scala I believe it is very important to describe the concept of mixins as well. It is something that Java developers are not used to but which is very flexible and that  people from Python/Ruby will be happy to see